desc/geometry/surface.py

```diff
@@ -357,6 +361,200 @@ def from_near_axis(cls, aspect_ratio, elongation, mirror_ratio, axis_Z, NFP=1):
+                    the base_surface evaluated at the input ``grid``
+                ``x`` : (``grid.num_nodes`` x 3) array of the position vectors on
+                    the base_surface evaluated at the input ``grid``
+                ``x_offset_surface`` : (``grid.num_nodes`` x 3) array of the
+                    position vectors on the offset surface, corresponding to the
+                    ``x`` points on the base_surface (i.e. the points to which the
+                    offset surface was fit)
+        info : tuple
+            2 element tuple containing residuals and number of iterations
+            for each point. Only returned if ``full_output`` is True
+
+        """
+        if grid is None:
+            grid = LinearGrid(
+                M=base_surface.M,
+                N=base_surface.N,
+                NFP=base_surface.NFP,
+                sym=base_surface.sym,
+            )
+        assert isinstance(
+            base_surface, FourierRZToroidalSurface
+        ), "base_surface must be a FourierRZToroidalSurface!"
+        M = base_surface.M if M is None else int(M)
+        N = base_surface.N if N is None else int(N)
+
+        Rbasis = base_surface.R_basis
+        Zbasis = base_surface.Z_basis
+
+        def n_and_r_jax(nodes):
+            R = Rbasis.evaluate(nodes) @ base_surface.R_lmn
+            Z = Zbasis.evaluate(nodes) @ base_surface.Z_lmn
+            R_t = Rbasis.evaluate(nodes, np.array([0, 1, 0])) @ base_surface.R_lmn
+            Z_t = Zbasis.evaluate(nodes, np.array([0, 1, 0])) @ base_surface.Z_lmn
+            R_z = Rbasis.evaluate(nodes, np.array([0, 0, 1])) @ base_surface.R_lmn
+            Z_z = Zbasis.evaluate(nodes, np.array([0, 0, 1])) @ base_surface.Z_lmn
+
+            phi = nodes[:, 2]
+            coords = jnp.stack([R, phi, Z], axis=1)
+            re = rpz2xyz(coords)
+            # NOTE: when generalized toroidal angle is implemented
+            # this must include omega
+            e_theta = jnp.array([R_t, jnp.zeros_like(R_t), Z_t]).T
+            e_zeta = jnp.array([R_z, R, Z_z]).T
+            e_theta_cross_e_zeta = cross(e_theta, e_zeta)
+            n = safediv(
+                e_theta_cross_e_zeta.T, jnp.linalg.norm(e_theta_cross_e_zeta, axis=1)
+            ).T
+            n = rpz2xyz_vec(n, phi=phi)
+            r_offset = re + offset * n
+            return n, re, r_offset
+
+        def fun_jax(zeta_hat, theta, zeta):
+            nodes = jnp.vstack((jnp.ones_like(theta), theta, zeta_hat)).T
+            n, r, r_offset = n_and_r_jax(nodes)
+            return jnp.arctan(r_offset[0, 1] / r_offset[0, 0]) - zeta
+
+        vecroot = jit(
+            vmap(
+                lambda x0, *p: root_scalar(
+                    fun_jax,
+                    x0,
+                    jac=None,
+                    args=p,
+                )
+            )
+        )
+        zetas, (res, niter) = vecroot(
+            grid.nodes[:, 2], grid.nodes[:, 1], grid.nodes[:, 2]
```

Comment:
Are the same values being passed in for both `zeta_hat` and `zeta` to the function `fun_jax`? If so, why bother naming them as separate variables in the function?

Reply:
They are separate variables int he function, one is the zeta on the offset surface, the other is the zeta on the base surface.

Initially we do not know the offset surface zeta, so we use the base surface zeta as an initial guess, hence why here we pass it in as the `x0` for the rootfind